### ISSUE TITLE
TL Detection: Train and implement neural network to detect traffic light bounding boxes and classify light color

### ISSUE THREAD

**Issue by edufford** [team-fusionx/CarND-Capstone#29]
Train and implement traffic light detector and classifier to publish /traffic_waypoint topic output when a red light is detected from the camera images.

Possible network candidates are:
1. MobileNet SSD
2. Faster RCNN
3. Consensus from NN 1 or 2 + classic SVM classifier on bounding box crop for redundancy

It looks like images publish at 10 Hz, so inference time should be optimized to be < 100ms on HW similar to Carla's:

* 31.4 GiB Memory
* Intel Core i7-6700K CPU @ 4 GHz x 8
* TITAN X Graphics

Also, submission file size limit is 2GB so frozen network files should be smaller than this.

**Comment by shrikondra** [comment]
Going through the requirements of the traffic light detector, we need to determine only the state of the light and not the exact location in the image.

I saw this post on the forum too, they say if SSD approach is used model size goes too big and created problems for submission.
https://discussions.udacity.com/t/how-to-upload-huge-light-classifier-model-into-github-for-submission/488648/2

So let us first not go through the SSD approach. 

I have listed down my checklist for classification task below.

- Extracted BOSCH Data - Done today

- Convert BOSCH, simulator, ros bag data for training and testing - 15th Mar
	- https://github.com/bosch-ros-pkg/bstld
	- To view all the data stats
	- save the data in 4 folders (red, green, yellow, none) for both test and train

- Use Mobilenet to do transfer learning - 18th Mar

- load the model in classfier script and return the State given the image - 20th Mar

- Retrain model if required / Use SSD- 24th Mar

**Comment by edufford** [comment]
I believe Udacity increased the file size submission limit to 2GB so we don’t have to worry about this restriction, but I agree that smaller and shallower is better for this project’s purpose.  We don’t have to do bounding box detection, but it may have some advantages for suppressing miss detection from glare scattered in the image (like glare and traffic light reflection in the car’s hood) and I was also thinking about cropping the bounding box image area to use as an additional redundant classification with traditional vision OpenCV processing or SVM to try a consensus based decision.  These are just ideas though so I’m ok with sticking to the simplest single full image classifier too if you would rather go that way.  Labeling sim training data will definitely be easier without bounding boxes.

One big challenge though is that we don’t have much data from Carla with the actual camera images to test with so it may be hard to have confidence in the final accuracy for the submission test.

**Comment by meenunat** [comment]
@edufford @shrikondra - sounds good. I have uploaded my ground truth pass (branch) - I see the waypoints - but I am currently debugging if the data is correct or not. Please let me know how I can help with the other tasks.

**Comment by edufford** [comment]
Initial training has been done for simulator classification with MobileNet and ResNet architectures.  ResNet seems to have good enough accuracy to proceed with integration testing.  Further model training on Udacity real camera images from bag data and Bosch dataset will continue from 3/30.  Moving this issue's milestone to 3rd Pass for the testing and tuning phase.

**Comment by edufford** [comment]
Trained model has been implemented in PR #89 and PR #91 that works for both simulation and real camera images from ROS bag.  Closing this issue.
